src/Npgsql/NpgsqlWriteBuffer.cs

```diff
@@ -98,15 +148,23 @@ public async Task Flush(bool async)
             try
             {
                 if (async)
-                    await Underlying.FlushAsync();
+                    await Underlying.FlushAsync(timeoutCt);
                 else
                     Underlying.Flush();
             }
+            catch (OperationCanceledException e)
+            {
+                throw Connector.Break(new NpgsqlException("Exception while flushing stream", new TimeoutException("Timeout during flushing attempt", e)));
+            }
             catch (Exception e)
             {
                 throw Connector.Break(new NpgsqlException("Exception while flushing stream", e));
             }
 
+            // Resetting cancellation token source, so we can use it again
+            if (async)
+                _timeoutCts.CancelAfter(-1);
+
```

Comment:
You can probably move this directly after the `FlushAsync()` call to avoid the second `if (async)` directly after the exception handlers.